docs/madr/decisions/056-policy-matching-with-real-resources.md

```diff
@@ -0,0 +1,185 @@
+        Value: foo
+    Conf:
+      connectionTimeout: 2s
+      http:
+        requestTimeout: 15s
+      idleTimeout: 20s
+    Origin: [ ... ]
+  - Subset:
+      - Key: kuma.io/service
+        Not: false
+        Value: bar
+    Conf:
+      connectionTimeout: 2s
+      http:
+        requestTimeout: 20s
+      idleTimeout: 20s
+    Origin: [ ... ]
+```
+
+This structure is used by both "from" and "to" policies plugins.
+Policy plugins use the `Rules` list to compute `Conf` for the known destination.
+In example, if we want to know the `Conf` for outbound with `kuma.io/service: bar` tags,
+we're going to iterate over `Rules` and check if our outbound belongs to the `Rules[].Subset`.
+
+The approach has a number of disadvantages:
+
+* even though it works well for "from" policies, using it for "to" policies doesn't feel natural
+* when referencing the real MeshHTTPRoute we had to use subset with artificial `__rule-matches-hash__` key
+* adding support for `MeshService` forces us to introduce another artificial key like `__resource-name__`
+* Inspect API with subsets is tricky when referencing real resources (like MeshHTTPRoute or MeshService)
+
+We need a new format that'd make things easier when it comes to referencing real resources in `to[]`.
+
+## Considered Options and Decision Outcome
+
+No changes for "from" policies, both policy matching algorithm and Inspect API stays the same.
+
+For "to" policies the algorithm is going to return the additional structure:
+
+```go
+type ToRules struct {
+    Rules         Rules 
+    ResourceRules map[UniqueResourceKey]ResourceRule // new field
+}
+
+type ResourceRule struct {
+    Resource core_model.ResourceMeta
+    Conf     interface{}
+    Origin   []core_model.ResourceMeta
+}
+```
+
+When targetRef has kind `MeshService` and uses `labels` we're going to produce multiple `ResourceRule` entities
+for each real `MeshService` that satisfies `labels`, i.e for the given `targetRef`:
+
+```yaml
+to:
+  - targetRef:
+      kind: MeshService
+      labels:
+        k8s.kuma.io/namespace: finance
+    conf: conf1
+```
+
+we are going to fetch all `MeshService` resources, filter them by `k8s.kuma.io/namespace: finance` and produce N rules:
+
+```yaml
+resourceRules:
+  - resource:
+      name: finance-backend.finance
+      mesh: mesh-1
+    conf: conf-1
+    origin: [...]
+  - resource:
+      name: finance-frontend.finance
+      mesh: mesh-1
+    conf: conf-1
+    origin: [...]
+  - resource:
+      name: finance-db.finance
+      mesh: mesh-1
+    conf: conf-1
+    origin: [...]
+```
+
+Policy plugins should implement a fallback logic when trying to get a conf for the given `MeshService{name: backend}`:
+
+1. Lookup `MeshService{name: backend}` in the new `ToRules.ResourceRules`
+2. If it exists then the real MeshService `backend` exists in the cluster and we can use `ResourceRule.Conf`
+3. If it's absent then fallback to legacy `ToRules.Rules` and get conf for `kuma.io/service: backend` tag
```

Comment:
Why do we need to fallback to legacy `ToRules.Rules`? How would this work? MeshService do not map directly to `kuma.io/service`, technically we have `.Status.Identities`, but I'd avoid using it for anything else than mTLS

Reply:
I don't think we neefd that fallback, when applying policy we know if we are looking for new ToRules or old subset, I think what is important is that we need to build ToRules and old Subset from policy, to handle old policies targeting old MeshService by name

Reply:
Ah right, it's not a fallback, we always know in the code when to use old `ToRules.Rules` and when to use new `ToRules.ResourceRules`